modern-publish-plugin.md

```diff
@@ -0,0 +1,211 @@
+# Modern Gradle Publish Plugin
+
+
+## Goals
+
+- less magic configuration, more explicit APIs
+- provide APIs for everything represented through properties
+- remove or deprecate APIs that are already provided by Gradle
+- continue providing a way to get started with minimal configuration
+
+## New Structure
+
+For compatibility purposes the current `com.vanniktech.maven.publish` plugin
+behavior is kept as it is, with the exception of APIs that we decide to remove. 
+Most of the goals are solved through a set of new plugins for specific 
+platforms. These plugins provide APIs that replace the magic setup mechanisms
+of the current plugin. To avoid having to maintain two implentations the old
+plugin will internally just apply the new plugins and use their public APIs. 
+Over time we can decide if `com.vanniktech.maven.publish` should be deprecated
+or removed.
+
+### Plugins
+
+Specific plugins for each supported project type:
+- `com.vanniktech.maven.publish.java`
+- `com.vanniktech.maven.publish.gradle`
+- `com.vanniktech.maven.publish.android`
+- `com.vanniktech.maven.publish.kotlin.js`
+- `com.vanniktech.maven.publish.kotlin.mpp`
+
+There are two reasons for going with individual plugins. 
+
+
+On top of that there will be a `com.vanniktech.maven.publish.base` plugin that 
+gets applied by all of these and the existing `com.vanniktech.maven.publish` as 
+described above.
+
+#### `com.vanniktech.maven.publish.base`
+
+- applies `maven-publish` Gradle plugin
+- provides the API that is shared for all project types
+- can be used for shared configurations using `plugins.withId(...)` in projects with mixed types
+
+```
+mavenPublish {
+  # applies signing plugin and tells Gradle to sign all publications
+  signAllPublications()
+  # adds the Sonatype release/snapshot repo as targets based on version
+  # uses Gradle credentials API for username/password
+  # automatically calls signAllPublications()
+  # sets nexus.baseUrl
+  publishToMavenCentral()
```

Comment:
Should this be used as an indicator by project plugins to automatically enable source/javadoc artifacts or is it just the magic `com.vanniktech.maven.publish` plugin that adds those automatically?

Reply:
Maybe it could be a parameter on this method. That way we get easy configurability without forcing it.

Reply:
Does not 99% of the people want to push to maven central?

Reply:
Yes, I still think it should be opt-in (so you need to something to enable it), but then it should be as effortless as possible. The reason for making it opt-in is to avoid having magic in the base plugins. I think having 2 parameters here to disable source/javadoc jar might be the best compromise. Just calling `publishToMavenCentral()` saves you from most config but it's still possible to enable it without automatically enabling javadoc/source jars (theoretically it's possible to have that for central).